### ISSUE TITLE
Dependency on GPL licensed version crate

### ISSUE THREAD

**Issue by est31** [matthiasbeyer/imag#1212]
Disclaimer: IANAL.

I've used [cargo-license](https://github.com/onur/cargo-license) to list the licenses of all dependencies of crates inside this workspace:

```
Apache-2.0/MIT (68): backtrace, backtrace-sys, bitflags, bitflags, boolinator, cc, cfg-if, chrono, chrono, dbus, dtoa, either, encode_unicode, env_logger, error-chain, error-chain, gcc, glob, idna, itertools, itoa, lazy_static, lazy_static, libc, log, log, notify-rust, num, num-integer, num-iter, num-traits, percent-encoding, pkg-config, quick-error, quote, rand, regex, regex-syntax, rust-crypto, rustc-demangle, rustc-serialize, semver, semver, semver-parser, serde, serde_derive, serde_derive_internals, serde_json, syn, synom, tempdir, tempfile, term, thread_local, time, toml, unicode-bidi, unicode-normalization, unicode-width, unicode-xid, unreachable, url, uuid, vec_map, winapi, winapi-i686-pc-windows-gnu, winapi-x86_64-pc-windows-gnu, xdg-basedir
BSD-3-Clause (4): fuchsia-zircon, fuchsia-zircon-sys, hoedown, prettytable-rs
CC0-1.0 (6): encoding-index-japanese, encoding-index-korean, encoding-index-simpchinese, encoding-index-singlebyte, encoding-index-tradchinese, encoding_index_tests
GPL-3.0 (1): version
MIT (27): ansi_term, atty, base64, block, clap, email, encoding, handlebars, iso8601, kernel32-sys, mac-notification-sys, malloc_buf, matches, nom, objc, objc-foundation, objc_id, redox_syscall, redox_termios, strsim, termion, termios, textwrap, vobject, void, winapi, winapi-build
MIT OR Apache-2.0 (1): rustc_version
MIT/Unlicense (8): aho-corasick, byteorder, csv, memchr, memchr, same-file, utf8-ranges, walkdir
MPL-2.0 (7): filters, is-match, kairos, pest, task-hookrs, toml-query, toml-query
N/A (50): imag, imag-annotate, imag-bookmark, imag-contact, imag-diagnostics, imag-diary, imag-edit, imag-gps, imag-grep, imag-habit, imag-init, imag-link, imag-log, imag-mail, imag-mv, imag-notes, imag-ref, imag-store, imag-tag, imag-timetrack, imag-todo, imag-view, libimagbookmark, libimagcontact, libimagdiary, libimagentryannotation, libimagentrycategory, libimagentrydatetime, libimagentryedit, libimagentryfilter, libimagentrygps, libimagentrylink, libimagentrylist, libimagentrymarkdown, libimagentryref, libimagentrytag, libimagentryutil, libimagentryview, libimagerror, libimaghabit, libimaginteraction, libimaglog, libimagmail, libimagnotes, libimagnotification, libimagrt, libimagstore, libimagtimetrack, libimagtimeui, libimagtodo
Unlicense (1): interactor
```

It seems to me that almost all licenses are compatible to the LGPL version 2.1 but one. I've feared that it would be more as solely Apache-2.0 licensed crates are popular on crates.io and they are incompatible with LGPL before version 3. The only crate that has a conflicting license is `version`. This is not very bad as the GPL is compatible with the LGPL, but the combined executable has to be distributed under GPL conditions. You should either make that clear somewhere, or use a different crate than `version`.

**Comment by matthiasbeyer** [comment]
Hi @est31,

thanks for your contribution. That's unfortunate! I filed an issue [in the repository for the version crate](https://github.com/nulldatamap/version/issues/5) for relicensing.

I also though about relicensing all the time (to MPL or GPLv2only) but because we have a few contributors already, that might be more complicated.

Do you think that relicensing would only require acknowledgement of the autors of the current master checkout or from _all_ contributors down the history of the project? Maybe another ping to @steveklabnik who might have some ideas on this (sorry if this annoys you Steve :smile: )?

**Comment by est31** [comment]
> Do you think that relicensing would only require acknowledgement of the autors of the current master checkout or from all contributors down the history of the project?

If you want to put anything under a new license you will need consent by all people who own copyright on your work. These are usually a subset of the contributors to your github repo. At least in Germany and I think in the US as well, trivial contributions don't count into copyright (in German copyright this is called "Schöpfungshöhe"), so some authors would be exempt. Whether something falls into copyright or counts as trivial contribution doesn't have an easy answer. Usually it is therefore most practical to simply ask all contributors for consent. If someone refuses or doesn't react you can still decide whether you want to judge their contribution as trivial or not.

Whether you need to ask every author throughout the entire history or only the authors of the latest version depends on whether the work from your previous authors still counts towards the copyright, so whether components of the latest version can be seen as derived work from the author's contributions.  I've also seen projects removing components created by people who haven't consented to some license change and rewriting them.

As I've said, I'm no lawyer, so only giving my best guesses here.

If you are doing a relicensing move, I would suggest you avoid GPLv2 or LGPLv2 because the two aren't compatible with the Apache2.0, or at least in the view of some lawyers. Other people don't see a problem, but better be on the safe side I guess :). Instead use the GPLv3 or LGPLv3, or a more permissive license like MPL or the Apache2.0/MIT combination. This gives you the freedom of integrating Apache 2.0 only licensed libraries in the future.

**Comment by steveklabnik** [comment]
> Do you think that relicensing would only require acknowledgement of the autors of the current master checkout or from all contributors down the history of the project?

I concur with @est31 here.

**Comment by matthiasbeyer** [comment]
Thanks for your suggestions. I do not like GPLv3 so MPL it would be, I guess. I would like to wait what https://github.com/nulldatamap/version/issues/5 results in and then we'll move forward.

Right now we have 24 authors (including me) in the imag codebase (`git log --format="%an %ae" | sort -u `, some might be duplicates because of faulty git configuration - [github only lists 19](https://github.com/matthiasbeyer/imag/graphs/contributors)), though only 13 of them have code in the latest master (7fdc71230ef4b312f0dc525a98b5908b3b1e037e) ... (`git ls-files | head -n 1 | while read line; do git blame --show-email "$line" | cut -d "(" -f 2 | cut -d " " -f 1 ; done | sort -u`).

At least two of the overall contributors have contributed a "trivial contribution" (only one line). So asking all would be the best idea, asking 22 (or, minus duplicates, 17) would be minimum.
Asking them via github sounds not right to me, because I want to move away from github in the long run anyways, so email it is.

**Comment by matthiasbeyer** [comment]
I created https://github.com/matthiasbeyer/imag/pull/1214 to fix this issue for now.

This is not a permanent fix, though. At least not as far as I understand it.

**Comment by matthiasbeyer** [comment]
As I merged #1214 I consider this solved.

I will open a new issue for relicensing.

**Comment by est31** [comment]
For the apache 2.0 issue, I've heard that one can also add an additional permission to the (L)GPL 2.0 license to make the two compatible. I don't know the details though.
